destination/destination.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return fmt.Errorf("failed to dial SSH: %w", err)
+	}
+
+	d.sftpClient, err = sftp.NewClient(d.sshClient)
+	if err != nil {
+		d.sshClient.Close()
+		return fmt.Errorf("failed to create SFTP client: %w", err)
+	}
+
+	_, err = d.sftpClient.Stat(d.config.DirectoryPath)
+	if err != nil {
+		return fmt.Errorf("remote path does not exist: %w", err)
+	}
+
+	return nil
+}
+
+func (d *Destination) Write(_ context.Context, records []opencdc.Record) (int, error) {
+	for i, record := range records {
+		chunked, ok := record.Metadata["is_chunked"]
+		if ok && chunked == "true" {
+			err := d.handleChunkedRecord(record)
+			if err != nil {
+				return i, err
+			}
+			continue
+		}
+
+		filename, ok := record.Metadata["filename"]
+		if !ok {
+			filename = string(record.Key.Bytes())
+		}
+
+		err := d.uploadFile(filename, record.Payload.After.Bytes())
+		if err != nil {
+			return i, err
+		}
+	}
+
+	return len(records), nil
+}
+
+func (d *Destination) handleChunkedRecord(record opencdc.Record) error {
+	index, ok := record.Metadata["chunk_index"]
+	if !ok {
+		return NewInvalidChunkError("chunk_index not found")
+	}
+
+	totalChunks, ok := record.Metadata["total_chunks"]
+	if !ok {
+		return NewInvalidChunkError("total_chunk not found")
+	}
+
+	hash, ok := record.Metadata["hash"]
+	if !ok {
+		return NewInvalidChunkError("hash not found")
+	}
+
+	var remoteFile *sftp.File
+	var err error
+	path := fmt.Sprintf("%s/%s.tmp", d.config.DirectoryPath, hash)
+	if index == "1" {
+		remoteFile, err = d.sftpClient.Create(path)
+		if err != nil {
+			return fmt.Errorf("failed to create remote file: %w", err)
+		}
+	} else {
+		remoteFile, err = d.sftpClient.OpenFile(path, os.O_WRONLY|os.O_APPEND)
+		if err != nil {
+			return fmt.Errorf("failed to open remote file: %w", err)
+		}
+	}
+
+	reader := bytes.NewReader(record.Payload.After.Bytes())
+	_, err = reader.WriteTo(remoteFile)
+	if err != nil {
+		return fmt.Errorf("failed to write content to remote file: %w", err)
+	}
+	remoteFile.Close()
+
+	if index == totalChunks {
+		filename, ok := record.Metadata["filename"]
+		if !ok {
+			filename = string(record.Key.Bytes())
+		}
+
+		err = d.sftpClient.Rename(path, fmt.Sprintf("%s/%s", d.config.DirectoryPath, filename))
+		if err != nil {
+			return fmt.Errorf("failed to rename remote file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (d *Destination) Teardown(ctx context.Context) error {
+	sdk.Logger(ctx).Info().Msg("Tearing down the SFTP Destination")
+
+	var errs []error
+	if d.sftpClient != nil {
+		if err := d.sftpClient.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("close SFTP client: %w", err))
+		}
+	}
+
+	if d.sshClient != nil {
+		if err := d.sshClient.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("close SSH client: %w", err))
+		}
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("error teardown: %w", errors.Join(errs...))
+	}
+
+	return nil
+}
+
+func (d *Destination) sshConfigAuth() (*ssh.ClientConfig, error) {
```

Comment:
Nitpick: we have a similar thing in the source, maybe we can extract this as a function that's available in the destination and source package as well.

Reply:
Done